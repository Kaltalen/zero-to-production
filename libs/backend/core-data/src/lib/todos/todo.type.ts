import { gql } from 'apollo-server-koa';

export const todoTypeDef = gql`
  type Todo {
    id: ID!
    userId: ID!
    title: String!
    description: String!
    completed: Boolean!
  }

  input NewTodoInput {
    userId: ID!
    title: String!
    description: String!
  }

  input UpdatedTodoInput {
    id: ID!
    userId: ID
    title: String
    description: String
    completed: Boolean
  }

  extend type Query {
    Todo(id: ID!): Todo!
    allTodos: [Todo]!
  }

  extend type Mutation {
    newTodo(input: NewTodoInput!): Todo!
    updateTodo(input: UpdatedTodoInput!): Todo!
    removeTodo(id: ID!): Todo!
  }
`;
