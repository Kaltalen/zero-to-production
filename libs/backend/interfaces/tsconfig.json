{
  "extends": "../../../tsconfig.json",
  "compilerOptions": {
    "module": "commonjs",
    "emitDeclarationOnly": true,
    "declaration": true,
    "composite": true,
    "rootDir": "./src",
    "types": ["node", "jest"]
  },
  "include": ["**/*.ts"]
}
