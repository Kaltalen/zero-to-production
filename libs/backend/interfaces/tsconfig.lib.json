{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "../../../dist/libs/backend/interfaces",
    "tsBuildInfoFile": "../../../dist/buildcache/backend-interfaces.tsbuildinfo",
    "types": []
  },
  "exclude": ["**/*.spec.ts"],
  "include": ["**/*.ts"]
}
