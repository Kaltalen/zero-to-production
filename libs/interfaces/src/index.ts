export * from './lib/user';
export * from './lib/todo';
