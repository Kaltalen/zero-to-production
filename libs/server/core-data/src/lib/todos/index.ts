export {
  ITodoDocument,
  ITodoModel,
  getTodoModel,
  createTodoModel,
  todoDbKey,
  todoSchema,
  todoNoteDbKey
} from './todo';
export { todoTypeDef } from './todo.type';
export { createTodosResolver } from './todo.resolver';
export { createTodosRouter } from './todo.router';

export {
  ITodoNoteDocument,
  ITodoNoteModel,
  getTodoNoteModel,
  createTodoNoteModel,
  todoNoteSchema
} from './notes';
