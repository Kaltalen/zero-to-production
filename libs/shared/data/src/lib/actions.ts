import { Action } from '@ngrx/store';

/**
 * Some state actions include a payload to update the new state.
 * Extend the Action type to include the payload.
 */
export interface ActionWithPayload<T> extends Action {
  readonly payload: T;
}
