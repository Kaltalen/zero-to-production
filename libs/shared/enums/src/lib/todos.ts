export enum TodoFilterStatus {
  All,
  Completed,
  InCompleted
}
