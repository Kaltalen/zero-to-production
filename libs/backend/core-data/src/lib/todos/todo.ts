import mongoose from 'mongoose';
import { defaultSchemaOptions } from '@uqt/backend/utils';
import { ITodo } from '@uqt/interfaces';
export { todoTypeDef } from './todo.type';

export const todoSchema = new mongoose.Schema<ITodo>(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      required: true,
      ref: 'user'
    },
    title: String,
    description: String,
    completed: {
      type: Boolean,
      required: true,
      default: false
    }
  },
  {
    ...defaultSchemaOptions
  }
);

export const Todo = mongoose.model<ITodoDocument, ITodoModel>(
  'todo',
  todoSchema
);

export interface ITodoDocument extends ITodo, mongoose.Document {
  id: string;
}

export interface ITodoModel extends mongoose.Model<ITodoDocument> {}

// export const todoTypeDefs = () => {
//   return todoTypeDef;

//   // TODO -> Cli Builder to copy non TS Files?
//   // return readFileSync(resolve(__dirname, './todo.graphql'), {
//   // encoding: 'utf-8'
//   // });
// };
