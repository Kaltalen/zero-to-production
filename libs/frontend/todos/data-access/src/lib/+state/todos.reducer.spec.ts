/**
 * A reducer is just a pure function, it takes inputs and return a
 * value when called because it is pure function we do not need to
 * configure any of angular testing module because it does not need
 * of angular configuration to operate. We can also just use snapshot
 * testing to test the output of the function and not need to use
 * assertion testing
 */
import { todosReducer, TodosEntityState } from './todos.reducer';
import * as TodoActions from './todos.actions';
import { EntityAdapter, createEntityAdapter } from '@ngrx/entity';
import { ITodo } from '@ngw/shared/interfaces';
import { TodoFilterStatus } from '@ngw/shared/enums';

describe('TodoReducer', () => {
  let adapter: EntityAdapter<ITodo>;

  const todo: ITodo = {
    id: '1',
    user: '1',
    title: 'some title',
    description: 'some description',
    completed: true
  };
  const initialState: TodosEntityState = {
    ids: [todo.id],
    entities: {
      [todo.id]: todo
    },
    selectedTodoId: null,
    statusFilter: TodoFilterStatus.InCompleted,
    searchFilter: null
  };

  beforeEach(() => {
    adapter = createEntityAdapter<ITodo>();
  });

  describe('undefined action', () => {
    it('should return the default state', () => {
      const action = {} as any;
      const result = todosReducer(undefined, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('LoadSuccess', () => {
    it('should add the todos to the todo state', () => {
      const todos: ITodo[] = [
        {
          id: '1',
          user: '1',
          title: 'some title',
          description: 'some description',
          completed: true
        },
        {
          id: '2',
          user: '1',
          title: 'another title',
          description: 'another description',
          completed: false
        }
      ];

      const action = TodoActions.loadTodosSuccess({ todos });
      const result = todosReducer(undefined, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('CreateSuccess', () => {
    it('should add a todo to the todo state', () => {
      const newTodo = {
        id: '2',
        user: '1',
        title: 'another title',
        description: 'another description',
        completed: false
      };

      const action = TodoActions.createTodoSuccess({ todo: newTodo });
      const result = todosReducer(initialState, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('UpdateSuccess', () => {
    it('should update the todo state', () => {
      const todo1 = {
        id: '1',
        user: '1',
        title: 'some title',
        description: 'some description',
        completed: false
      };
      const todo2 = { ...todo1, id: '2', title: 'anotherTodo' };

      const state: TodosEntityState = {
        ids: [todo1.id, todo2.id],
        entities: {
          [todo1.id]: todo1,
          [todo2.id]: todo2
        },
        selectedTodoId: null,
        statusFilter: TodoFilterStatus.InCompleted,
        searchFilter: null
      };

      const updateTodo = {
        id: '2',
        completed: true
      } as ITodo;

      const action = TodoActions.updateTodoSuccess({
        todo: { id: updateTodo.id, changes: updateTodo }
      });
      const result = todosReducer(state, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('DeleteSuccess', () => {
    it('should delete the todo from the sate', () => {
      const todo1 = {
        id: '1',
        user: '1',
        title: 'some title',
        description: 'some description',
        completed: true
      };
      const todo2 = { ...todo1, id: '2', title: 'anotherTodo' };

      const state: TodosEntityState = {
        ids: [todo1.id, todo2.id],
        entities: {
          [todo1.id]: todo1,
          [todo2.id]: todo2
        },
        selectedTodoId: null,
        statusFilter: TodoFilterStatus.InCompleted,
        searchFilter: null
      };

      const todoToDelete = {
        id: '2'
      } as ITodo;

      const action = TodoActions.deleteTodoSuccess(todoToDelete);
      const result = todosReducer(state, action);
      expect(result).toMatchSnapshot();
    });
  });
});
