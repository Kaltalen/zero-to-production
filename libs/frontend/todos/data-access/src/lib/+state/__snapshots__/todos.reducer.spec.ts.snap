// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoReducer CreateSuccess should add a todo to the todo state 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": true,
      "description": "some description",
      "id": "1",
      "title": "some title",
      "user": "1",
    },
    "2": Object {
      "completed": false,
      "description": "another description",
      "id": "2",
      "title": "another title",
      "user": "1",
    },
  },
  "ids": Array [
    "1",
    "2",
  ],
  "searchFilter": null,
  "selectedTodoId": null,
  "statusFilter": 2,
}
`;

exports[`TodoReducer DeleteSuccess should delete the todo from the sate 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": true,
      "description": "some description",
      "id": "1",
      "title": "some title",
      "user": "1",
    },
  },
  "ids": Array [
    "1",
  ],
  "searchFilter": null,
  "selectedTodoId": null,
  "statusFilter": 2,
}
`;

exports[`TodoReducer LoadSuccess should add the todos to the todo state 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": true,
      "description": "some description",
      "id": "1",
      "title": "some title",
      "user": "1",
    },
    "2": Object {
      "completed": false,
      "description": "another description",
      "id": "2",
      "title": "another title",
      "user": "1",
    },
  },
  "ids": Array [
    "1",
    "2",
  ],
  "searchFilter": null,
  "selectedTodoId": null,
  "statusFilter": 2,
}
`;

exports[`TodoReducer UpdateSuccess should update the todo state 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "description": "some description",
      "id": "1",
      "title": "some title",
      "user": "1",
    },
    "2": Object {
      "completed": true,
      "description": "some description",
      "id": "2",
      "title": "anotherTodo",
      "user": "1",
    },
  },
  "ids": Array [
    "1",
    "2",
  ],
  "searchFilter": null,
  "selectedTodoId": null,
  "statusFilter": 2,
}
`;

exports[`TodoReducer undefined action should return the default state 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
  "searchFilter": null,
  "selectedTodoId": null,
  "statusFilter": 2,
}
`;
